ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/JmsXAResourceRecoveryHelper.java

```diff
@@ -177,4 +268,11 @@ private void disconnect() throws XAException {
         }
     }
 
+    private void assertIfRecoveryScanWasStarted() throws XAException {
+        if (delegate == null) {
+            LOGGER.warn("XAResource.recover has to be called first with XAResource.TMSTARTRSCAN argument");
+            throw new XAException(XAException.XAER_PROTO);
```

Comment:
Are we sure XAER_PROTO is correct.  Calls to this class are completely under the control of the recovery module so it would be a coding error on our behalf if we do not call recover correctly or fail to handle the error if the start scan call throws an exception. There is an argument that the correct code is XAER_INVAL since the definition says that XAER_PROTO should only be used if one of the transitions in the state table is violated:

"An RM must ensure that each calling thread makes xa_ calls in a legal sequence,
and must return [XAER_PROTO] or other suitable error if the caller violates the
state tables (see Chapter 6)."

But since calls to this class are private to our recovery system perhaps we are being too picky.